### PR TITLE
Add AfmoeForCausalLM support

### PR DESCRIPTION
**Pull request by bartowski1182**
Adds support for upcoming AfmoeForCausalLM

Tokenizer is public ahead of model launch to avoid breaking conversion code

*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

### CODE REVIEW

File: convert_hf_to_gguf.py [hunk 1]
```diff
@@ -2457,6 +2460,100 @@ def set_gguf_parameters(self):
...
+        if (n_shared_experts := self.hparams.get("num_shared_experts")) is not None:
+            self.gguf_writer.add_expert_shared_count(n_shared_experts)
+        if (moe_intermediate_size := self.hparams.get("moe_intermediate_size")) is not None:
+            self.gguf_writer.add_expert_feed_forward_length(moe_intermediate_size)
+        if (n_dense_layers := self.hparams.get("num_dense_layers")) is not None:
+            self.gguf_writer.add_leading_dense_block_count(n_dense_layers)
+
+        # Gating function (sigmoid)
+        if (score_func := self.hparams.get("score_func")) is not None and score_func == "sigmoid":
+            self.gguf_writer.add_expert_gating_func(gguf.ExpertGatingFuncType.SIGMOID)
```
Comment (CISC) [inline convert_hf_to_gguf.py]: Even though it's only using `sigmoid` now it's preferable that you check and set the ones we support:
https://github.com/ggml-org/llama.cpp/blob/13aaafe633f1b01aa72c04ce519d8901894cf5f1/convert_hf_to_gguf.py#L8662-L8667

File: convert_hf_to_gguf.py [hunk 2]
```diff
@@ -2457,6 +2460,100 @@ def set_gguf_parameters(self):
...
+        elif ".pre_mlp_layernorm." in name and bid is not None:
+            return [(self.format_tensor_name(gguf.MODEL_TENSOR.FFN_PRE_NORM, bid), data_torch)]
+
+        # Post FFN norm
+        elif ".post_mlp_layernorm." in name and bid is not None:
+            return [(self.format_tensor_name(gguf.MODEL_TENSOR.FFN_POST_NORM, bid), data_torch)]
+
+        # Map router
+        elif ".mlp.router.gate." in name and bid is not None:
+            return [(self.format_tensor_name(gguf.MODEL_TENSOR.FFN_GATE_INP, bid), data_torch)]
```
Comment (CISC) [inline convert_hf_to_gguf.py]: Surely all of these can be added to `tensor_mapping.py` instead (in fact it looks like you did for `pre_mlp_layernorm`)?

File: models/ggml-vocab-afmoe.gguf
Comment (CISC) [inline models/ggml-vocab-afmoe.gguf:1]: Delete

File: models/ggml-vocab-afmoe.gguf.inp
Comment (CISC) [inline models/ggml-vocab-afmoe.gguf.inp:1]: Delete

File: models/ggml-vocab-afmoe.gguf.out
Comment (CISC) [inline models/ggml-vocab-afmoe.gguf.out:1]: Delete

File: src/llama-model.cpp [hunk 3]
```diff
@@ -695,6 +696,39 @@ void llama_model::load_hparams(llama_model_loader & ml) {
                     default: type = LLM_TYPE_UNKNOWN;
                 }
             } break;
+        case LLM_ARCH_AFMOE:
+            {
+                ml.get_key(LLM_KV_ATTENTION_LAYERNORM_RMS_EPS, hparams.f_norm_rms_eps);
+                ml.get_key(LLM_KV_LEADING_DENSE_BLOCK_COUNT,   hparams.n_layer_dense_lead);
+                ml.get_key(LLM_KV_EXPERT_FEED_FORWARD_LENGTH,  hparams.n_ff_exp);
+                ml.get_key(LLM_KV_EXPERT_COUNT,                hparams.n_expert);
+                ml.get_key(LLM_KV_EXPERT_USED_COUNT,           hparams.n_expert_used);
```
Comment (CISC) [inline src/llama-model.cpp]:
```suggestion
```

File: src/models/afmoe.cpp [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            // MoE layer with sigmoid routing, normalization, and scaling
+            ggml_tensor * moe_out = build_moe_ffn(cur,
+                    model.layers[il].ffn_gate_inp,
+                    model.layers[il].ffn_up_exps,
+                    model.layers[il].ffn_gate_exps,
+                    model.layers[il].ffn_down_exps,
+                    model.layers[il].ffn_exp_probs_b,
+                    n_expert, n_expert_used,
+                    LLM_FFN_SILU,
+                    hparams.expert_weights_norm != 0,      // norm_w (route_norm=True)
```
Comment (CISC) [inline src/models/afmoe.cpp]:
```suggestion
                    hparams.expert_weights_norm,           // norm_w (route_norm=True)
```

File: src/models/afmoe.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    res->t_embd = cur;
+
+    // lm_head
+    cur = build_lora_mm(model.output, cur);
+    cb(cur, "result_output", -1);
+    res->t_logits = cur;
+
+    ggml_build_forward_expand(gf, cur);
+}
```
Comment (CISC) [inline src/models/afmoe.cpp]:
```suggestion
}

```

File: models/ggml-vocab-afmoe.gguf
Comment (CISC) [reply models/ggml-vocab-afmoe.gguf:1]: I assume you've tested this with `test-tokenizer-0`, we don't need the files.

File: convert_hf_to_gguf.py (hunk 2 above)
Comment (ngxson) [reply convert_hf_to_gguf.py]: Same question, seems like this can be specified in `tensor_mapping.py`

Comment (bartowski1182) [reply convert_hf_to_gguf.py]: Totally valid, wasn't sure where the line was between adding new stuff to `tensor_mapping` versus mapping in the conversion code, will move those over! 

(the pre_mlp_layernorm I did after my initial implementation and that's when I noticed I should consider the tensor mapping)

Comment (bartowski1182) [reply convert_hf_to_gguf.py]: (also awkwardly most of the mappings exist, I forgot to look for them after we renamed a bunch of our tensors, so I won't be adding much to `tensor_mapping.py`, thanks!)

File: convert_hf_to_gguf.py [hunk 6]
```diff
@@ -2457,6 +2460,82 @@ def set_gguf_parameters(self):
...
+            self.gguf_writer.add_expert_count(n_experts)
+        if (n_shared_experts := self.hparams.get("num_shared_experts")) is not None:
+            self.gguf_writer.add_expert_shared_count(n_shared_experts)
+        if (moe_intermediate_size := self.hparams.get("moe_intermediate_size")) is not None:
+            self.gguf_writer.add_expert_feed_forward_length(moe_intermediate_size)
+        if (n_dense_layers := self.hparams.get("num_dense_layers")) is not None:
+            self.gguf_writer.add_leading_dense_block_count(n_dense_layers)
+
+        # Expert Gating Function
+        score_func = self.hparams.get("score_func", "sigmoid")
```
Comment (bartowski1182) [inline convert_hf_to_gguf.py]: Is it acceptable to assign a default of "sigmoid" here @CISC ?

File: models/ggml-vocab-afmoe.gguf
Comment (bartowski1182) [reply models/ggml-vocab-afmoe.gguf:1]: `test-tokenizer-0.sh` doesn't want to run, is there any documentation on how to run it?

I get `./tests/test-tokenizer-0.sh: line 26: ./tests/test-tokenizer-0: No such file or directory`

I assume because of the `make` error I see as I start the test:

```
Makefile:6: *** Build system changed:
 The Makefile build has been replaced by CMake.

 For build instructions see:
 https://github.com/ggml-org/llama.cpp/blob/master/docs/build.md

.  Stop.
```

File: src/models/afmoe.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+            // Q/K normalization
+            Qcur = build_norm(Qcur, model.layers[il].attn_q_norm, NULL, LLM_NORM_RMS, il);
+            Kcur = build_norm(Kcur, model.layers[il].attn_k_norm, NULL, LLM_NORM_RMS, il);
+            cb(Qcur, "Qcur_normed", il);
+            cb(Kcur, "Kcur_normed", il);
+
+            // RoPE only for sliding_attention layers (every 4th layer is full_attention)
+            // layer_types[i] = "sliding_attention" if (i+1) % global_attn_every_n_layers != 0
+            bool is_sliding = ((il + 1) % 4) != 0;  // global_attn_every_n_layers = 4
```
Comment (ngxson) [inline src/models/afmoe.cpp]: It seems like this model support sliding window. In such case, I would recommend follow the same pattern with llama 4, example can be found in `models/llama-iswa.cpp`

To avoid hard-coding the number `4` here, you can set it on hparams, again have a look on `LLM_ARCH_LLAMA4` where they use `hparams.set_swa_pattern(4);`

File: models/ggml-vocab-afmoe.gguf
Comment (CISC) [reply models/ggml-vocab-afmoe.gguf:1]: Use the executable, not the script (you need to build `llama.cpp`).

File: src/models/afmoe.cpp (hunk 7 above)
Comment (bartowski1182) [reply src/models/afmoe.cpp]: seems funny to me to use `n_no_rope_layer_step`, which is hardcoded as `4` in `llama-hparams.h`, and then `hparams.set_swa_pattern(4);` is also hardcoded at 4, but for now I've fixed as suggested :)

File: convert_hf_to_gguf.py (hunk 6 above)
Comment (CISC) [reply convert_hf_to_gguf.py]: Well, if you expect `score_func` not to be defined, and considering you default to `sigmoid` here:
https://github.com/ggml-org/llama.cpp/blob/34dc2a377d85ae0668d816b4e902d3abf476dc37/src/llama-model.cpp#L719-L722
..perhaps you should only raise an error if `score_func` is not `None`
https://github.com/ggml-org/llama.cpp/blob/1b9558f8352f955a2ac2f7c0fcb03487df2e3818/convert_hf_to_gguf.py#L2486-L2487

File: models/ggml-vocab-afmoe.gguf
Comment (bartowski1182) [reply models/ggml-vocab-afmoe.gguf:1]: ah thanks that explains it..

turns out that after some internal tokenizer updates, I was in fact not using the correct regex, so thanks for pointing it out !

unfortunately it's gonna take some unicode.cpp additions, so working on that now..

File: convert_hf_to_gguf_update.py [hunk 8]
```diff
@@ -139,6 +139,7 @@ class TOKENIZER_TYPE(IntEnum):
     {"name": "lfm2",             "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/LiquidAI/LFM2-Tokenizer"},
     {"name": "exaone4",          "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/LGAI-EXAONE/EXAONE-4.0-32B", },
     {"name": "mellum",           "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/JetBrains/Mellum-4b-base", },
+    {"name": "afmoe",            "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/arcee-ai/Trinity-Tokenizer", },
```
Comment (CISC) [inline convert_hf_to_gguf_update.py:142]: This is no good, [config.json](https://huggingface.co/arcee-ai/Trinity-Tokenizer/blob/main/config.json) is empty and can't be used to make the vocab GGUF.

Comment (bartowski1182) [reply convert_hf_to_gguf_update.py:142]: yeah i noticed this and will push a fix that works, one sec

Comment (bartowski1182) [reply convert_hf_to_gguf_update.py:142]: sorry about that, pushed a minimum viable one that worked for me locally, hope that's acceptable!

File: src/unicode.cpp [hunk 9]
```diff
@@ -742,6 +816,9 @@ static std::vector<size_t> unicode_regex_split_custom(const std::string & text,
     } else if (regex_expr == "\\p{Han}+") {
         // K2's first pattern - handle all K2 patterns together
         bpe_offsets = unicode_regex_split_custom_kimi_k2(text, offsets);
+    } else if (regex_expr == "\\p{Nd}+") {
```
Comment (bartowski1182) [inline src/unicode.cpp]: my main concern is that this feels pretty generic and i'm worried someone else will hit on it accidentally, is that a reasonable concern or should I leave it as-is? this part of the code starts to go over my head a bit, but it does currently work..

File: convert_hf_to_gguf.py (hunk 6 above)
Comment (bartowski1182) [reply convert_hf_to_gguf.py]: I expect it to be defined but wanted to be more robust, maybe to a fault..

I could either set it to NONE on purpose to be caught in llama-model.cpp, or remove my defaulting and let it break if it breaks 🤷

File: src/llama-vocab.cpp [hunk 10]
```diff
@@ -445,8 +445,9 @@ struct llm_tokenizer_bpe : llm_tokenizer {
                 break;
             case LLAMA_VOCAB_PRE_TYPE_AFMOE:
                 regex_exprs = {
-                    // Digits in groups of 1-3
-                    "\\p{N}{1,3}",
+                    // Digit handling - uses custom implementation in unicode.cpp
+                    // Groups digits with leading 1-2 based on total length modulo 3
+                    "\\p{Nd}+",
```
Comment (CISC) [inline src/llama-vocab.cpp]: That's not strictly the meaning of this character class, though I see why you'd want to do this. Can this perhaps be replaced with a fictional class instead? How is this handled in `transformers`?

File: src/unicode.cpp (hunk 9 above)
Comment (CISC) [reply src/unicode.cpp]: Yeah, not a fan, see below.

File: convert_hf_to_gguf_update.py (hunk 8 above)
Comment (CISC) [reply convert_hf_to_gguf_update.py:142]: Perfect

File: convert_hf_to_gguf.py [hunk 11]
```diff
@@ -2457,6 +2460,82 @@ def set_gguf_parameters(self):
...
+        if score_func == "sigmoid":
+            self.gguf_writer.add_expert_gating_func(gguf.ExpertGatingFuncType.SIGMOID)
+        elif score_func == "softmax":
+            self.gguf_writer.add_expert_gating_func(gguf.ExpertGatingFuncType.SOFTMAX)
+        else:
+            raise ValueError(f"Unsupported score_function value: {score_func}")
```
Comment (CISC) [inline convert_hf_to_gguf.py]:
```suggestion
        score_func = self.hparams.get("score_func")
        if score_func == "sigmoid":
            self.gguf_writer.add_expert_gating_func(gguf.ExpertGatingFuncType.SIGMOID)
        elif score_func == "softmax":
            self.gguf_writer.add_expert_gating_func(gguf.ExpertGatingFuncType.SOFTMAX)
        elif score_func is not None:
            raise ValueError(f"Unsupported score_function value: {score_func}")
```

File: convert_hf_to_gguf.py (hunk 6 above)
Comment (CISC) [reply convert_hf_to_gguf.py]: See suggestion.

File: convert_hf_to_gguf.py [hunk 12]
```diff
@@ -2457,6 +2460,82 @@ def set_gguf_parameters(self):
...
+                    merged_name = f"model.layers.{bid}.mlp.experts.{w_name}.weight"
+                    new_name = self.map_tensor_name(merged_name)
+                    tensors.append((new_name, data_torch))
+
+                return tensors
+            else:
+                return []
+
+
+        if name.endswith(".expert_bias"):
```
Comment (CISC) [inline convert_hf_to_gguf.py:2532]:
```suggestion
                return []

        if name.endswith(".expert_bias"):
```

File: src/llama-vocab.cpp (hunk 10 above)
Comment (bartowski1182) [reply src/llama-vocab.cpp]: in transformers it's handled with a complex regex that doesn't seem to be supported in std::regex, the lookaheads were causing problems that don't seem supported

Definitely not opposed to replacing with a fictional class, would even prefer it, just wasn't positive how fictional I can get with it so I'll take a look

Comment (bartowski1182) [reply src/llama-vocab.cpp]: ahh i see that if it matches it doesn't even attempt to parse it, so I could use `\\p{AFMoE}` which will almost certainly never collide!

Comment (CISC) [reply src/llama-vocab.cpp]: Yes, though perhaps even more descriptive like `AFMoE_digits`.

Comment (bartowski1182) [reply src/llama-vocab.cpp]: done

**Review by CISC — APPROVED** [review]

File: src/models/afmoe.cpp [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+    ggml_tensor * cur;
+    ggml_tensor * inpL;
+
+    inpL = build_inp_embd(model.tok_embd);
+
+    // MuP scaling: embeddings * sqrt(hidden_size)
+    // mup_enabled = true, hidden_size = 1024, scale = 32.0
+    inpL = ggml_scale(ctx0, inpL, sqrtf(float(n_embd)));
+    cb(inpL, "inp_embd_scaled", -1);
```
Comment (ngxson) [inline src/models/afmoe.cpp:15]: not very important to fix right now, but if the model supports multimodal in the future, you may need to skip scaling if the input is non-text:

https://github.com/ggml-org/llama.cpp/blob/45c6ef7307e8ab383a8b6557f8ed8d237b26c452/src/models/gemma3-iswa.cpp#L11-L15

**Review by ngxson — APPROVED** [review]

File: src/models/afmoe.cpp (hunk 13 above)
Comment (bartowski1182) [reply src/models/afmoe.cpp:15]: Interesting, good to know thanks !

**Comment by CISC** [comment]
@bartowski1182 @ggerganov Merging in a little while unless you have anything more to add.
